psh/tools/basic.py

```diff
@@ -14,43 +14,70 @@
 #
 
 import pexpect
+from trunner.config import CURRENT_TARGET, DEVICE_TARGETS
 
+# psh prompt and end of line regex
+PROMPT = r'(\r+)\x1b\[0J' + r'\(psh\)% '
+EOL = r'(\r+)\n'
 
-def assert_expected(p, cmd, expected, msg=''):
-    p.sendline(cmd)
-    p.expect_exact(cmd)
-    assert p.expect([expected, pexpect.TIMEOUT]) == 0, msg
 
+class Psh:
```

Comment:
Same problem as below in this case. I see it like there is one pexpect process object in test: `p`. The `psh` object is created based on pexpect process `p`. We don't need to know about that pexpect process in psh test, we only want to communicate with psh, so there is psh object, which we use. We don't need to think about asserting prompt and so on. We only have methods to passing commands and executing programs with specified output via psh. That's why I think assertions should also be in psh class (second reason is about missing unwanted output, when it's split, please see the example below).

Reply:
> We don't need to know about that pexpect process in psh test, we only want to communicate with psh, so there is psh object, which we use. We don't need to think about asserting prompt and so on. We only have methods to passing commands and executing programs with specified output via psh. That's why I think assertions should also be in psh class (second reason is about missing unwanted output, when it's split, please see the example below).

Now we have to care about `psh` instance of Psh class and `p` instance of pyexpect process. If we implement psh interface as a module we only use `p` argument.

We still won't need to think about asserting prompt and so on. I want to exec command and do not care about prompt assertion? OK i use `psh.exec_cmd(p, cmd)`. I want to parse output by myself? Ok, I can use, for example, `psh.input(p, cmd)`, `p.read(some_number_of_bytes)`, `psh.assert_raw_prompt(p)`. This can be helpful when I write test for `ps` and I would like to iterate line by line over my output. I have nice interface to make sure that my command (echoed input) was correctly asserted and I've read valid prompt.

> And that's why I can't import module here (I need p argument from harness function in psh object)

If we implement psh as a module, there is no need to create any object. We pass pyexpect process instance to the functions directly.

Reply:
That's right, we have to care about psh instance and p instance, but once in the beginning of the test. If we are staying with passing `p`, I agree the idea with psh module is absolutely fine. I will try to do it in the way You suggest. I am not sure about the need for add `psh.input` method, we can use `p.sendline` for example.